### PR TITLE
[ARCTIC-925]Add serializer for spilled map and resolve life cycle of spilled map

### PR DESCRIPTION
**Pull request by nicochen**
## Why are the changes needed?
fix: #925 
1. Currently, spilled maps use java native serializers, but in the case of record comparison, the serialized bytes of records with the same primary key may be different, which leads to inaccurate comparison results, so it is a spilled map Customized serializer to ensure that the serialized byte of data with the same primary key is the same.
2. At present, the spilled map used in the MOR main process does not handle the life cycle of the spilled map well.

## Brief change log
1. Add com.netease.arctic.utils.map.Serializer for spilled map
2. Reconstruct all DeleteFilter to resolve life cycle of map

## How was this patch tested?
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

  - Does this pull request introduces a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

### CODE REVIEW

File: core/src/main/java/com/netease/arctic/iceberg/CombinedDeleteFilter.java [hunk 1]
```diff
@@ -416,4 +430,51 @@ public RecordWithLsn recordCopy() {
       return this;
     }
   }
+
+  private class CloseableIterableWithOtherCloseable<T> implements CloseableIterable<T> {
```
Comment (majin1102) [inline core/src/main/java/com/netease/arctic/iceberg/CombinedDeleteFilter.java:434]: could it be named as CloseableIterableWrapper?

File: core/src/main/java/com/netease/arctic/iceberg/optimize/DeleteFilter.java [hunk 2]
```diff
@@ -285,4 +301,54 @@ private static Schema fileProjection(Schema tableSchema, Schema requestedSchema,
 
     return new Schema(columns);
   }
+
+
+  private class CloseableIterableWithOtherCloseable<T> implements CloseableIterable<T> {
```
Comment (majin1102) [inline core/src/main/java/com/netease/arctic/iceberg/optimize/DeleteFilter.java:306]: This is the same implementation as inner class of CombinedDeleteFilter, could it be reused?

File: core/src/main/java/com/netease/arctic/iceberg/CombinedDeleteFilter.java (hunk 1 above)
Comment (nicochen) [reply core/src/main/java/com/netease/arctic/iceberg/CombinedDeleteFilter.java:434]: ok

File: core/src/main/java/com/netease/arctic/iceberg/optimize/DeleteFilter.java (hunk 2 above)
Comment (nicochen) [reply core/src/main/java/com/netease/arctic/iceberg/optimize/DeleteFilter.java:306]: There are some differences between the two implementations，but reuse is not impossible。I will make some changes

File: core/src/main/java/com/netease/arctic/io/reader/ArcticDeleteFilter.java [hunk 3]
```diff
@@ -444,4 +458,51 @@ private Schema fileProjection(
...
+      return new CloseableIteratorWithOtherCloseable<>(closeableIterator);
+    }
+
+    @Override
+    public void close() throws IOException {
+      inner.close();
+    }
+  }
+
+  private class CloseableIteratorWithOtherCloseable<T> implements CloseableIterator<T> {
```
Comment (majin1102) [inline core/src/main/java/com/netease/arctic/io/reader/ArcticDeleteFilter.java:482]: the same as above

File: core/src/main/java/com/netease/arctic/utils/map/JavaSerializer.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+ */
+
+package com.netease.arctic.utils.map;
+
+import com.netease.arctic.utils.SerializationUtils;
+
+import java.io.IOException;
+import java.io.Serializable;
+
+public class JavaSerializer<T extends Serializable> implements Serializer<T> {
```
Comment (majin1102) [inline core/src/main/java/com/netease/arctic/utils/map/JavaSerializer.java:26]: shall we put this into SerializationUtils together?

File: core/src/main/java/com/netease/arctic/utils/map/RocksDBBackend.java [hunk 5]
```diff
@@ -47,37 +45,65 @@
...
+  }
+
   private Map<String, ColumnFamilyHandle> handlesMap = new HashMap<>();
   private Map<String, ColumnFamilyDescriptor> descriptorMap = new HashMap<>();
   private RocksDB rocksDB;
   private boolean closed = false;
   private final String rocksDBBasePath;
   private long totalBytesWritten;
+  private Serializer<K> keySerializer;
+  private Serializer<T> valueSerializer;
```
Comment (majin1102) [inline core/src/main/java/com/netease/arctic/utils/map/RocksDBBackend.java:87]: we could put Serializers into Map implementations and use byte[] for put and get

File: core/src/main/java/com/netease/arctic/utils/map/Serializer.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netease.arctic.utils.map;
+
+public interface Serializer<T> {
```
Comment (majin1102) [inline core/src/main/java/com/netease/arctic/utils/map/Serializer.java:21]: suggest put these Serializer interface and implementation into SerializationUtils

File: core/src/main/java/com/netease/arctic/utils/map/StructLikeWrapperSerializer.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+import com.netease.arctic.iceberg.optimize.StructLikeCopy;
+import com.netease.arctic.iceberg.optimize.StructLikeWrapper;
+import com.netease.arctic.iceberg.optimize.StructLikeWrapperFactory;
+import com.netease.arctic.utils.SerializationUtils;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.types.Types;
+
+import java.io.IOException;
+
+public class StructLikeWrapperSerializer implements Serializer<StructLikeWrapper> {
```
Comment (majin1102) [inline core/src/main/java/com/netease/arctic/utils/map/StructLikeWrapperSerializer.java:30]: move this to SerilizationUtils

File: core/src/main/java/com/netease/arctic/utils/map/StructLikeFactory.java [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netease.arctic.utils.map;
+
+import com.netease.arctic.utils.StructLikeSet;
+import org.apache.iceberg.types.Types;
+
+public class StructLikeFactory {
```
Comment (majin1102) [inline core/src/main/java/com/netease/arctic/utils/map/StructLikeFactory.java:24]: MapIdentifier should not be shared

**Review by majin1102 — CHANGES_REQUESTED** [review]
check some reusable classes

File: core/src/main/java/com/netease/arctic/io/reader/GenericIcebergDataReader.java [hunk 9]
```diff
@@ -56,7 +56,7 @@ public GenericIcebergDataReader(
       boolean caseSensitive,
       BiFunction<Type, Object, Object> convertConstant,
       boolean reuseContainer,
-      StructLikeFactory structLikeFactory) {
+      StructLikeCollections structLikeFactory) {
```
Comment (majin1102) [inline core/src/main/java/com/netease/arctic/io/reader/GenericIcebergDataReader.java:59]: parameter name is not fixed yet

File: hive/src/main/java/com/netease/arctic/hive/io/reader/AdaptHiveBaseArcticDataReader.java [hunk 10]
```diff
@@ -49,7 +49,7 @@ public AdaptHiveBaseArcticDataReader(
       BiFunction<Type, Object, Object> convertConstant,
       Set<DataTreeNode> sourceNodes,
       boolean reuseContainer,
-      StructLikeFactory structLikeFactory) {
+      StructLikeCollections structLikeFactory) {
```
Comment (majin1102) [inline hive/src/main/java/com/netease/arctic/hive/io/reader/AdaptHiveBaseArcticDataReader.java:52]: parameter name is not fixed yet

**Review by majin1102 — CHANGES_REQUESTED** [review]
Parameter names of structLikeFactory has not been fixed

**Review by majin1102 — APPROVED** [review]
LGTM! thanks for contributing

**Pull request merged**